Adaptors/RabbitMQ/src/PullQueueStorage.cs

```diff
@@ -60,132 +54,54 @@ public override Task<HealthCheckResult> Check(HealthCheckTag tag)
-        // Cancellation token will be set by the pulling function.
-        queueMessageHandlers_.Enqueue(new QueueMessageHandler(ConnectionRabbit.Channel!,
-                                                              eventArgs,
-                                                              message,
-                                                              cancellationToken));
-        return Task.CompletedTask;
-      }
-
-      consumer.Received += Subscriber;
-
-
-      ConnectionRabbit.Channel!.BasicConsume(pullQueue.QueueName,
-                                             false,
-                                             consumer);
-      IsInitialized = true;
-    }
+                        "x-queue-mode", "lazy" // queue will try to move messages to disk as early as practically possible
+                      },
+                    };
+
+    ConnectionRabbit.Channel!.QueueDeclare(Options!.PartitionId,
+                                           false, /* to survive broker restart */
+                                           false, /* used by multiple connections */
+                                           false, /* not deleted when last consumer unsubscribes (if it has had one) */
+                                           queueArgs);
+    IsInitialized = true;
   }
 
-  public IAsyncEnumerable<IQueueMessageHandler> PullMessagesAsync(int               nbMessages,
-                                                                  CancellationToken cancellationToken = default)
+  public async IAsyncEnumerable<IQueueMessageHandler> PullMessagesAsync(int               nbMessages,
+                                                                        CancellationToken cancellationToken = default)
   {
+    var nbPulledMessage = 0;
+
     if (!IsInitialized)
     {
       throw new InvalidOperationException($"{nameof(PullQueueStorage)} should be initialized before calling this method.");
     }
 
-    cancellationToken.ThrowIfCancellationRequested();
-
-    // using ToAsyncEnumerable avoids using a needless async operator
-    return Enumerable.Repeat(0,
-                             nbMessages)
-                     .Select(_ =>
-                             {
-                               IQueueMessageHandler? qmh;
-                               do
-                               {
-                                 cancellationToken.ThrowIfCancellationRequested();
-                               } while (!queueMessageHandlers_!.TryDequeue(out qmh));
-
-                               return qmh;
-                             })
-                     .ToAsyncEnumerable();
+    while (nbPulledMessage < nbMessages)
+    {
+      cancellationToken.ThrowIfCancellationRequested();
+
+      var message = ConnectionRabbit.Channel!.BasicGet(Options.PartitionId,
+                                                       false);
+
+      if (message is null)
+      {
+        continue;
```

Comment:
If there is no message, you should sleep for a bit, or at the very least yield.
```suggestion
        await Task.Delay(100)
                  .ConfigureAwait(false);
        continue;
```